### ISSUE TITLE
corruption when linking receipts

### ISSUE THREAD

**Issue by andreva7** [clovett/MyMoney.Net#76]
I link a bunch of receipts, save, come back later, more later. Save, link more. Then I discover some links are confused - corrupted.
Some items have a little clippy, but no image in them anymore.
Some other items that should have a single receipt ends up with multiple receipts, from other transactions.

Not sure how to debug this ...

One such transaction that has four receipts have IDs 83, 83.1, 83.2, 83.3 in the attachment directory. They are however unrelated receipts.

**Comment by lovettchris** [comment]
that happened to me once when I found out my onedrive folder was not synching properly.  Are you synching or backing up these folders? Then check the backups?  Also renaming an account and not saving that rename could also cause problems.

**Comment by andreva7** [comment]
This is happening with the school PTA account. There is no account renaming that happens there.
I do keep my data on on drive, but the mymoney file and the attachments are both on the same onedrive. It seems unlikely that one syncs and the other one does not, but I guess not impossible.
It makes sense to me that some receipts could be missing if there is a onedrive problem. I am not sure how that would explain multiple separate receipts, linked to separate transactions, at some point all show up together under another transaction.

-Andre

**Comment by lovettchris** [comment]
So OneDrive is in the picture, same as in my case.  In my case what was happening is the money database file was being reverted to an older version, which meant those transactions disappeared, but the receipts remained.  Then when I added another transaction id "reused" that transaction id, picking up the receipt, then if this happens again and again you could end up with a transaction id with multiple receipts.

So what I ended up doing was making the "master" copy of everything to be on a hard drive NOT in a OneDrive folder, and then I use robocopy to copy FROM there to a onedrive folder.  But I never edit anything on the onedrive folder using MyMoney.  Then everything after that has been solid.

**Comment by andreva7** [comment]
This does not appear to be the same at first glance.
No transactions has ever disappeared for my database.
I also have a transactions that say there is an attachment (because I linked it in the past, and remember doing do), but when I click on clippy, the receipt window shows nothing. That receipt shows up attached to some other transaction somewhere.

**Comment by andreva7** [comment]
I am looking at the DB using a SQLLite DB browser. At first glance I cannot find any column that links the receipts to the transaction... Do you know where that is stored off the top of your head?

**Comment by andreva7** [comment]
Figured out some things, using the SQL browser and without reading the code, 'cause I am a bit lazy.
The file name is the actual transaction ID.  You scan the directory for all filenames to determine what has a receipt.
When I click on a clippy and saw nothing it must be because the file contents was not synced (yet) to the machine. Could the code could be resilient to that somehow?

I looked more into a specific case where one transaction had 4 unrelated receipts attached to it:
The receipt file names were 83 83.1 83.2 83.3 and transaction 83 showed all four files in the mymoney receipt browser 
Looking at the receipt amounts and pictures, I tracked down that these receipts really belonged to transactions 830, 831, 836, 839
Quite a coincidence ...

**Comment by lovettchris** [comment]
yes the file name of the receipt literally contains the transaction id and if there's more than one receipt for a transaction they are numbers id.1, id.2, id.3, etc.

so transaction 83 is really old if you are up to 830, 831, 836, 839.  I would worry about how badly scrambled things are, and then stop using onedrive for a while and see if the problem goes away ?

**Comment by andreva7** [comment]
I have 2000+ transactions in the account right now. 
I am in the process of scanning the past 10 years of receipts that I got in a big box from the past treasurer and attaching them to all these transactions so I can get rid of all this paper.
 
I am reviewing everything I have linked up to now and cleaning up the bad linking. I found three instances of bad linking. The two previous ones I had cleaned up using the UI, so I have no forensics about it.
This transaction (83) is the only one I have more details about. If I find another I will investigate the IDs again.

It just seems there may be an off-by-one string handling issue in some error path.

**Comment by lovettchris** [comment]
I will double check the string parsing but I think it literally just splits the name on "." and treats the first part as a transaction id and the second part as an attachment index, so I have no ideas right now on how that could have gotten confused.

**Comment by andreva7** [comment]
All this happened with old, existing transactions.

I scan a batch of 10-15 receipts. For each receipt, I find the matching transaction and link it by sniping the receipt image from the scan and pasting it in the receipt UI. It was just somewhat random that I happened to link receipts for transactions 830, 831, 836, 839. But receipts are somewhat organized by date, so it makes sense they would be on transactions close in number.

If there was a bug, it would be as part of the insertion of the receipt on an existing transaction. The addition did work because the jpeg got saved in the attachments folder. If there was a bug it would have to be on the file name generation.

**Comment by lovettchris** [comment]
Oh wait, I think I have found a bug.  The following code is looking for `t.Id + "*.*"` but there is no "." separator after t.Id which means yes, indeed transaction `83` can pickup the attachment for 830 exactly as you are suggesting.  So you have helped me debug this, thanks so much!  I haven't seen this because my transaction numbers are no so high that it would be unlikely to happen (68565.jpg :-). I'm working on a fix.

```csharp
        public List<string> GetAttachments(string path, Transaction t)
        {
            List<string> files = new List<string>();
            if (!string.IsNullOrEmpty(path) && Directory.Exists(path) && t.Account != null && !string.IsNullOrEmpty(t.Account.Name))
            {
                string accountDirectory = Path.Combine(path, NativeMethods.GetValidFileName(t.Account.Name));
                if (Directory.Exists(accountDirectory))
                {
                    files.AddRange(Directory.GetFiles(accountDirectory, t.Id + "*.*"));                    
                }
            }
            return files;
        }
```

**Comment by clovett** [comment]
Fix pushed cc84e40041505c29c94d8b30ae7cfd4d53f4ac22

**Comment by clovett** [comment]
I also pushed this fix https://github.com/clovett/MyMoney.Net/issues/77 and published new version 2.0.1.2.  This fix I also improves the page up/page down scrolling performance in large accounts.  Let me know if you notice any improvement :-)
